src/transform.rs

```diff
@@ -0,0 +1,365 @@
+        }
+    }
+}
+impl PostVisit for ArrayValidation {
+    fn post_visit<V: VisitorMut>(&mut self, visitor: &mut V) {
+        // SingleOrVec doesn't give a way to get a mutable iterator, so this was
+        // the easiest way I found to iterate over it.
+        if let Some(items) = &mut self.items {
+            match items {
+                schemars::schema::SingleOrVec::Single(x) => x.post_visit(visitor),
+                schemars::schema::SingleOrVec::Vec(vec) => {
+                    for x in vec {
+                        x.post_visit(visitor)
+                    }
+                }
+            }
+        }
+        let children = self.additional_items.iter_mut().chain(&mut self.contains);
+        for child in children {
+            child.post_visit(visitor);
+        }
+    }
+}
+
+impl PostVisit for ObjectValidation {
+    fn post_visit<V: VisitorMut>(&mut self, visitor: &mut V) {
+        let children = self
+            .properties
+            .values_mut()
+            .chain(self.pattern_properties.values_mut())
+            .chain(self.additional_properties.as_deref_mut())
+            .chain(self.property_names.as_deref_mut());
+
+        for child in children {
+            child.post_visit(visitor);
+        }
+    }
+}
+
+struct UnionType<'a> {
+    root: &'a RootSchema,
+}
+
+impl VisitorMut for UnionType<'_> {
+    fn visit_object(&mut self, obj: &mut SchemaObject) {
+        if obj.instance_type.is_some() {
+            return;
+        }
+
+        if let Some(subschemas) = &mut obj.subschemas {
+            match (&subschemas.any_of, &subschemas.one_of) {
+                (None, Some(any_of)) | (Some(any_of), None) => {
+                    let Some(plain_types) = any_of
+                        .iter()
+                        .map(|s| plain_schema_types(s, self.root))
+                        .collect::<Option<Vec<_>>>()
+                    else {
+                        return;
+                    };
+
+                    let Some(mut plain_types) = plain_types
+                        .into_iter()
+                        .map(|tys| match tys {
+                            SingleOrVec::Single(t) => Some(*t),
+                            SingleOrVec::Vec(_) => None,
+                        })
+                        .collect::<Option<Vec<_>>>()
+                    else {
+                        return;
+                    };
+
+                    // Technically, if this is `one_of` then we should filter out any repeated types.
+                    plain_types.sort();
+                    plain_types.dedup();
+
+                    subschemas.any_of = None;
+                    subschemas.one_of = None;
+
+                    obj.instance_type = match plain_types.len() {
+                        0 => None,
+                        1 => Some(SingleOrVec::Single(Box::new(plain_types[0]))),
+                        _ => Some(SingleOrVec::Vec(plain_types)),
+                    };
```

Comment:
Without the Rust LSP I'm not sure of the underlying types, but I wonder if we could make this match better by matching on `plain_types.as_slice()` instead